examples/uniform-values/src/main.rs

```diff
@@ -0,0 +1,338 @@
+                WindowEvent::CloseRequested => {
+                    control_flow.set_exit();
+                }
+                WindowEvent::KeyboardInput { input, .. } => {
+                    let state_mut = state.as_mut().unwrap();
+                    let wgpu_context_ref = wgpu_context.as_ref().unwrap();
+                    if let Some(virtual_keycode) = input.virtual_keycode {
+                        // (7a)
+                        match virtual_keycode {
+                            VirtualKeyCode::Escape => control_flow.set_exit(),
+                            VirtualKeyCode::Up => state_mut.translate_view(1, 1),
+                            VirtualKeyCode::Down => state_mut.translate_view(-1, 1),
+                            VirtualKeyCode::Left => state_mut.translate_view(-1, 0),
+                            VirtualKeyCode::Right => state_mut.translate_view(1, 0),
+                            VirtualKeyCode::U => state_mut.max_iterations += 3.0,
+                            VirtualKeyCode::D => state_mut.max_iterations -= 3.0,
+                            _ => {}
+                        }
+                        wgpu_context_ref.window.request_redraw();
+                    }
+                }
+                WindowEvent::MouseWheel { delta, .. } => {
+                    let change = match delta {
+                        winit::event::MouseScrollDelta::LineDelta(_, vertical) => vertical,
+                        winit::event::MouseScrollDelta::PixelDelta(pos) => pos.y as f32 / 20.0,
+                    };
+                    let state_mut = state.as_mut().unwrap();
+                    let wgpu_context_ref = wgpu_context.as_ref().unwrap();
+                    // (7b)
+                    state_mut.zoom(change);
+                    wgpu_context_ref.window.request_redraw();
+                }
+                WindowEvent::Resized(new_size) => {
+                    let wgpu_context_mut = wgpu_context.as_mut().unwrap();
+                    wgpu_context_mut.resize(new_size);
+                    wgpu_context_mut.window.request_redraw();
+                }
+                _ => {}
+            },
+            Event::RedrawRequested(id) if id == main_window_id => {
+                let wgpu_context_ref = wgpu_context.as_ref().unwrap();
+                let state_ref = state.as_ref().unwrap();
+                let frame = wgpu_context_ref.surface.get_current_texture().unwrap();
+                let view = frame
+                    .texture
+                    .create_view(&wgpu::TextureViewDescriptor::default());
+
+                // (8)
+                wgpu_context_ref.queue.write_buffer(
+                    &wgpu_context_ref.uniform_buffer,
+                    0,
+                    // Ew gross I know. We have to do this to circumvent bytemuck's requirement
+                    // that AppState implement POD (Plain Old Data).
```

Comment:
Maybe worth expanding what "this" is. Reading this I focused on `bytemuck::bytes_of` and missed the fact that the members are extracted into an array of floats. Also it's something every wgpu user have to deal with and can easily lead to UB so I think it's worth expanding into more words.
How about:

We need to copy our app state into an array of bytes. This is typically done by copying the structure directly using `bytemuck::cast_slice` which, to avoid undefined behavior, requires that the struct implement the `POD` trait. Here, we instead first extract the data into an array of floats which fulfills `bytemuck`'s requirements and cast it into a byte slice using `bytemuck::bytes_of`.

Reply:
Does [291f11c](https://github.com/gfx-rs/wgpu/pull/3885/commits/291f11cb223ac5e9f441f1e26bac220efedb5010) work well or is it too much?

Reply:
I personally don't mind long comments, however:

> The problem is that there are a lot of things you need to be able to guarantee about how _Rust_ lays out the struct in memory

You can't rely on rust's struct layout at all so I think it would be more fair to say instead:

> The danger is that the struct _must_ be `#[repr(C)]`, its layout _must_ carefully match that of the structure in the shader, and _must_ not contain any padding.

And then,

> Generally, translating structs to bytes using bytemuck and POD should be avoided.

It's a strong stance, I don't entirely disagree but I don't fully agree either and I don't expect everyone in the project shares the opinion.

How about:

> Mistakes are easily made and not caught at build time, so translating structs to bytes using bytemuck and POD should be done very carefully.

Reply:
>You can't rely on rust's struct layout at all so I think it would be more fair to say instead...

I changed it to be more strongly-worded about the difficulty of predicting Rust's structure memory layout. It can be predicted but the rules are complex.

As for your being more soft on people implementing POD, I really strongly believe that it should be emphasized that 90% of users don't know nearly enough about memory internals to be able to reasonably judge when POD can be safely implemented. As you mention, mistakes are easily made and are not caught at build time and in addition, bugs resulting from guessing wrong can be absolutely infuriating to diagnose. I really believe that newer users should just skip learning how to implement POD and instead spend their time learning how to serialize structs in a WGSL-compatible way as even if you know there will be a padding incongruity, the best you can do it just not implement POD but then what? How do you get it to WGSL safely if you don't know how to serialize anyways?

I'm on and off writing a WGPU book and have a whole chapter planned for how to do that serialization.

See [0a1f2eb](https://github.com/gfx-rs/wgpu/pull/3885/commits/0a1f2ebc8a5582b10a0e2d2fd863c3b9fc28f9e6).

Reply:
I brought this up at the wgpu maintainer's call and the resolution is that if we want to encourage an opinionated way to transfer data to the gpu it should be via the encase crate.

Reply:
Cool! What exactly would this mean for this PR? Should I add a mention of encase or ought I reword significant parts of that comment?

Reply:
For this PR this would mean replace `bytemuck::bytes_of` and the float slice with something that uses encase (I haven't used encase so I hope what I'm saying makes sense, I'm relaying what was said during the call).

> I really believe that newer users should just skip learning how to implement POD 

I see where you are coming from but you take this stance to a degree that I and other maintainers don't entirely agree with. It's best if the docs that are associated in this repository keep things at a less opinionated level (a level that reflects wgpu's maintainers stance). That level is to explain the requirements and risks, mention encase as a solution for avoiding UB hazards and let users decide what is right for their level of expertise with that information.

For the wording:

> Rust representations of structs are not only incompatible but also complicated to predict

In the general case you simply can't predict it. Let's just say:

> Rust makes no guarantees about the layout of structs unless they are annotated with a specific representation such as #[repr(C)]. 


> It's not only safer but often necessary to implement translations to bytes manually

This is debatable (I at least don't think that the necessary part is true). Let's say something along the lines of: 

> Since this is error prone, we suggest using the encase crate for writing structures into buffers.

Reply:
>For this PR this would mean replace bytemuck::bytes_of and the float slice with something that uses encase (I haven't used encase so I hope what I'm saying makes sense, I'm relaying what was said during the call).

Sounds good. I'll start work on this when I can.

>I see where you are coming from but you take this stance to a degree that I and other maintainers don't entirely agree with. It's best if the docs that are associated in this repository keep things at a less opinionated level (a level that reflects wgpu's maintainers stance). That level is to explain the requirements and risks, mention encase as a solution for avoiding UB hazards and let users decide what is right for their level of expertise with that information.

I think this is very fair, especially the part about it being too opinionated. And also since it seems to be the position of the maintainers in general (since you brought this up at the maintainer's call), I'm absolutely going to honor that.

>In the general case you simply can't predict it. Let's just say:

Although Rust's layout of structs is (to my understanding) theoretically predictable, I can definitely get behind that it's not reasonable to try due to the lack of official guarantees and I'm down with the wording you provide.

Also, I think it definitely needed clarifying anyways that by "manually implementing WGSL serialization", I meant to say 'something that wasn't trying to blindly hammer a square, C-compatible peg into a star-shaped WGPU hole' and didn't mean to exclude things like the encase crate. I can 100% see how my poor wording confused though.

Reply:
No worries we are definitely going deep in a very nuanced topic, and you did a good job of explaining your position. Thanks for your understanding.

Reply:
Ok what do you think of the commenting on and implementation of [c5e95f5](https://github.com/gfx-rs/wgpu/pull/3885/commits/c5e95f5ed7749ffeeeb181a151b3fb3047b77359)?